src/asm/memmove.S

```diff
@@ -8,5 +8,73 @@ SECTION .TEXT
 
 _memmove:
 memmove:
-	mov eax, 5
+    ; Function arguments:
+    ; - RDI:    Dest pointer
+    ; - RSI:    Src pointer
+    ; - RDX:    Number of bytes to copy
+
+    ; Register layout
+    ; - rdi:    Destination address
+    ; - rsi:    Source address
+    ; - rax:    Scratch
+    ; - bh:     Whether we are reversing
+    ; - rcx:    Counter
+    ; - rdx:    Raw n. bytes to copy
+
+    push rdi            ; Save the dest addr
+    push rbx            ; We need another register
+
+    pop rbx
+    pop rax
+ret
+
+    mov rcx, rdx
+
+    cmp rdi, rsi        ; Check if rdi > rsi, if so, we decrement addrs
+    jna not_above
+
+    std                 ; Set direction flag
+    mov bh, 1           ; Set local direction flag
+    add rdi, rcx
+    sub rdi, 1          ; rdi += rcx - 1
+    add rsi, rcx
+    sub rsi, 1          ; rsi += rcx - 1
+
+not_above:
+    cld                 ; Unset direction flag
+
+qwords_check:
+    shr rcx, 3          ; rcx = rcx >> 3
+    cmp rcx, 0          
+    jne bytes           ; If <64B to copy, skip qw section
+
+qwords:
+    cmp bh, 1           ; First, determine if we are descending or ascending.
+    jne qwords_fwds     ; Skip padding and unpadding process?
+
+qwords_in_rev:
+    sub rdi, 3          ; rdi -= 3
+    sub rsi, 3          ; rsi -= 3
```

Comment:
Not gonna lie I don't get why this offset of 3 is here -- it's not scaling ofc, but I don't get why a constant offset would matter when you're calling a rep-prefixed instr anyways

Reply:
When we call `movsq <DEST>, <SRC>` the QWORD is written to the 4-byte block _beginning_ at `<DEST>`--since we've already added a 1B offset in the reverse case, we need to add 3 more bytes of offset so we don't segfault over the end of our block.

Reply:
This has just reminded me that QWORDS is eight bytes. Fix incoming